internal/telemetry/client.go

```diff
@@ -389,48 +390,114 @@ func (c *Client) newRequest(t RequestType) *Request {
 			LanguageVersion: runtime.Version(),
 		},
 		Host: Host{
-			Hostname:    hostname,
-			ContainerID: internal.ContainerID(),
-			OS:          getOSName(),
-			OSVersion:   getOSVersion(),
+			Hostname:     hostname,
+			OS:           getOSName(),
+			OSVersion:    getOSVersion(),
+			Architecture: runtime.GOARCH,
+			// TODO (lievan): getting kernel name, release, version TBD
 		},
 	}
+
+	header := &http.Header{
+		"Content-Type":               {"application/json"},
+		"DD-Telemetry-API-Version":   {"v2"},
+		"DD-Telemetry-Request-Type":  {string(t)},
+		"DD-Client-Library-Language": {"go"},
+		"DD-Client-Library-Version":  {version.Tag},
+		"DD-Agent-Env":               {c.Env},
+		"DD-Agent-Hostname":          {hostname},
+		"Datadog-Container-ID":       {internal.ContainerID()},
+	}
+	if c.URL == getAgentlessURL() {
+		header.Set("DD-API-KEY", c.APIKey)
+	}
+	client := c.Client
+	if client == nil {
+		client = defaultHTTPClient
+	}
+	return &Request{Body: body,
+		Header:          header,
+		HTTPClient:      client,
+		URL:             c.URL,
+		TelemetryClient: c}
 }
 
-// submit posts a telemetry request to the backend
-func (c *Client) submit(r *Request) error {
-	b, err := json.Marshal(r)
-	if err != nil {
-		return err
+// submit sends a telemetry request
+func (r *Request) submit() error {
+	if r.TelemetryClient == nil {
+		return fmt.Errorf("all telemetry requests must be associated with a telemetry client")
 	}
+	retry, err := r.trySubmit()
+	if retry {
+		// retry telemetry submissions in instances where the telemetry client has trouble
+		// connecting with the agent
+		r.TelemetryClient.log("telemetry submission failed, retrying with agentless: %s", err)
+		r.URL = getAgentlessURL()
+		r.Header.Set("DD-API-KEY", defaultAPIKey())
+		_, err := r.trySubmit()
+		if err != nil {
+			r.TelemetryClient.log("retrying with agentless telemetry failed: %s", err)
+		}
+	}
+	return err
+}
 
-	req, err := http.NewRequest(http.MethodPost, c.URL, bytes.NewReader(b))
+// agentlessRetry determines if we should retry a failed a request with
+// by submitting to the agentless endpoint
+func agentlessRetry(req *Request, resp *http.Response, err error) bool {
+	if req.URL == getAgentlessURL() {
+		// no need to retry with agentless endpoint if it already failed
+		return false
+	}
 	if err != nil {
-		return err
+		// we didn't get a response which might signal a connectivity problem with
+		// agent - retry with agentless
+		return true
 	}
-	req.Header = http.Header{
-		"Content-Type":              {"application/json"},
-		"DD-Telemetry-API-Version":  {"v1"},
-		"DD-Telemetry-Request-Type": {string(r.RequestType)},
+	// Do not retry with the following status codes:
+	// 400 - client side error
+	// 429 - too many requests
+	// TODO - add more
```

Comment:
Do you want to do this in a follow-up PR, or make a Jira ticket with this info so we don't lose track of it?